thunder/core/transform_common.py

```diff
@@ -463,133 +463,289 @@ def is_functionalizable(bsym: BoundSymbol) -> bool:
-                orig_args and len(orig_args) == 1,
-                lambda: f"{orig_args = } found for {copy_return = }",
-            )
-
-            copy_src = bsym.flat_proxy_args[0]
-            new_copy_dst = orig_args[0]
-            with tracectx(trace):
-                new_copy_out: TensorProxy = prims.copy_.meta(copy_src, new_copy_dst)
-                new_copy_bsym = prims.copy_.bind(copy_src, new_copy_dst, output=new_copy_out)
-            swap_map[variableify(copy_src)] = new_copy_out
-            additional_copy_bsyms.append(new_copy_bsym)
-            new_copy_dst_to_copy_bsyms[variableify(new_copy_dst)] = [new_copy_bsym]
-
-    narrowed_swap_map: dict[VariableInterface, TensorProxy] = {}
-    copy_bsyms = []
-    for var_t in copy_dst_to_copy_bsym:
-        if var_t in new_copy_dst_to_copy_bsyms:
-            bsyms = new_copy_dst_to_copy_bsyms[var_t]
-            copy_bsyms.extend(bsyms)
-            var_t = variableify(bsym.flat_proxy_args[0])
-            narrowed_swap_map[var_t] = swap_map[var_t]
-        else:
-            copy_bsyms.append(copy_dst_to_copy_bsym[var_t])
-    for var_t in (variableify(a) for a in view_args):
-        if var_t not in copy_dst_to_copy_bsym and var_t in new_copy_dst_to_copy_bsyms:
-            copy_bsyms.extend(new_copy_dst_to_copy_bsyms[var_t])
+                new_bsyms.append(reshape_bsym)
+            new_copy_return = prims.copy_.meta(reshaped_copy_src, new_copy_dst)
+            new_copy_bsym = prims.copy_.bind(reshaped_copy_src, new_copy_dst, output=new_copy_return)
+            new_bsyms.append(new_copy_bsym)
 
-    return copy_bsyms, narrowed_swap_map
+            if (var_t := variableify(new_copy_dst)) not in new_dst_to_copy_bsym:
+                new_dst_to_copy_bsym[var_t] = idx, new_bsyms
+            else:
+                cur_copy_bsym_idx, _ = new_dst_to_copy_bsym[var_t]
+                if cur_copy_bsym_idx < idx:
+                    new_dst_to_copy_bsym[var_t] = idx, new_bsyms
+
+    required_copy_bsyms = [None for _ in range(len(flat_args))]
+    for var_t in dst_to_copy_bsym:
+        idx = trace_args_set[unvariableify(var_t)]
+        cur_copy_bsym_idx, cur_copy_bsym = dst_to_copy_bsym[var_t]
+        if var_t not in new_dst_to_copy_bsym:
+            required_copy_bsyms[idx] = cur_copy_bsym
+        else:
+            new_copy_bsym_idx, new_copy_bsyms = new_dst_to_copy_bsym[var_t]
+            if cur_copy_bsym_idx < new_copy_bsym_idx:
+                required_copy_bsyms[idx] = new_copy_bsyms
+            else:
+                required_copy_bsyms[idx] = cur_copy_bsym
+    for var_t in new_dst_to_copy_bsym:
+        if var_t not in dst_to_copy_bsym:
+            _, bsyms = new_dst_to_copy_bsym[var_t]
+            required_copy_bsyms[trace_args_set[unvariableify(var_t)]] = bsyms
+
+    sorted_copy_bsyms: list[BoundSymbol] = []
+    for bsym_or_list in required_copy_bsyms:
+        if bsym_or_list is None:
+            continue
+        if isinstance(bsym_or_list, list):
```

Comment:
So this is/can be a list for the case when a reshape is necessary? Or are there cases when there are multiple `copy_` (I would expect not to). Could it make sense to make it always a list (and sometimes a one-element one)?

Reply:
> So this is/can be a list for the case when a reshape is necessary?

Correct.

> Could it make sense to make it always a list (and sometimes a one-element one)?

I don't have a strong preference

Reply:
3197cf9e37203e257b91cad892605bc7c19e6c94 switched to list

Reply:
Thank you!